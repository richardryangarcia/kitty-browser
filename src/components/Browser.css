.header {
  text-align: center;
  margin-top: 3em;
}

.search-btns {
  text-align: center;
  margin-top: 1em;
}

button {
  margin-right: 1em;
}

.card {
  margin-top: 1em;
  line-height: 4em;
  margin: 2em 4em;
  border: none;
  -webkit-box-shadow: 5px 5px 15px rgba(0, 0, 0, 0.4);
  -moz-box-shadow: 5px 5px 15px rgba(0, 0, 0, 0.4);
}

.kitty-details {
  padding: 5em 5em;
  font-size: 12px;
}

.kitty-img {
  text-align: center;
}
