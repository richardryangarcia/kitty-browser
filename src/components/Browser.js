import React, { Component } from "react";
import { object } from "prop-types";
import Web3 from "web3";
import KittyCoreABI from "../contracts/KittyCoreABI.json";
import { CONTRACT_NAME, CONTRACT_ADDRESS } from "../config";
import { formatDate } from "../utils/date";

class Browser extends Component {
  constructor(props, context) {
    super(props);
    this.state = { kittyId: 0, birthTime: null, generation: null, genes: null };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.contracts = context.drizzle.contracts;
  }

  componentDidMount() {
    const web3 = new Web3(window.web3.currentProvider);
    // Initialize the contract instance
    const kittyContract = new web3.eth.Contract(
      KittyCoreABI, // import the contracts's ABI and use it here
      CONTRACT_ADDRESS
    );
    // Add the contract to the drizzle store
    this.context.drizzle.addContract({
      contractName: CONTRACT_NAME,
      web3Contract: kittyContract
    });
  }

  handleChange(event) {
    this.setState({
      kittyId: event.target.value,
      birthTime: null,
      generation: null,
      genes: null
    });
  }

  async handleSubmit(event) {
    event && event.preventDefault();
    let response;
    try {
      response = await this.contracts.CryptoKitties.methods
        .getKitty(this.state.kittyId)
        .call();
      this.setState({
        birthTime: response.birthTime,
        generation: response.generation,
        genes: response.genes
      });
    } catch (e) {
      this.setState({
        birthTime: null,
        generation: null,
        genes: null
      });
    }
  }

  render() {
    const { birthTime, genes, generation, kittyId } = this.state;
    let formattedBday = birthTime && formatDate(birthTime);

    return (
      <div className="browser">
        <h1>Kitty Browser</h1>

        <form onSubmit={this.handleSubmit}>
          <h3>Kitty ID:</h3>
          <input
            type="text"
            value={this.state.kittyId}
            onChange={this.handleChange}
          />
          <button type="submit">Submit</button>
        </form>

        <button
          onClick={() => {
            this.setState({ kittyId: Math.round(Math.random() * 1713872) });
            this.handleSubmit();
          }}
          style={{ backgroundColor: "green", color: "white" }}
        >
          Random Kitty
        </button>

        {birthTime && (
          <div>
            <h3>Genes</h3>
            <p>{genes}</p>
            <h3>Generation</h3>
            <p>{generation}</p>
            <h3>Birth Time</h3>
            <p>{formattedBday}</p>
            <img
              src={`https://img.cryptokitties.co/0x06012c8cf97bead5deae237070f9587f8e7a266d/${kittyId}.svg`}
              style={{ height: "500px", width: "500px" }}
              alt="kitty"
            />
          </div>
        )}
      </div>
    );
  }
}

Browser.contextTypes = {
  drizzle: object
};

export default Browser;
