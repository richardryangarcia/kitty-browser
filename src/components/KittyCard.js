import React from "react";
import { formatDate } from "../utils/date";
import { Card } from "react-bootstrap";

export const KittyCard = props => {
  const { birthTime, genes, generation, kittyId } = props;
  let formattedBday = birthTime && formatDate(birthTime);

  return (
    <Card>
      <Card.Body>
        <div className="kitty-details">
          <h5>Kitty ID</h5>
          <p>{kittyId}</p>
          <h5>Genes</h5>
          <p>{genes}</p>
          <h5>Generation</h5>
          <p>{generation}</p>
          <h5>Birth Time</h5>
          <p>{formattedBday}</p>
        </div>
        <div className="kitty-img">
          <img
            src={`https://img.cryptokitties.co/0x06012c8cf97bead5deae237070f9587f8e7a266d/${kittyId}.svg`}
            style={{ height: "500px", width: "500px" }}
            alt="kitty"
          />
        </div>
      </Card.Body>
    </Card>
  );
};
