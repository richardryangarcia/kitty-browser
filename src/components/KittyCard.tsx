import React from "react";
import { formatDate } from "../utils/date";
import { Card } from "react-bootstrap";

type KittyCardProps = {
  birthTime?: Date;
  genes: number;
  generation: number;
  kittyId: string;
};

export const KittyCard: React.FC<KittyCardProps> = ({
  birthTime,
  genes,
  generation,
  kittyId
}) => {
  let formattedBday = birthTime && formatDate(birthTime);

  return (
    <Card>
      <Card.Body>
        <div className="kitty-details">
          <h5>Kitty ID</h5>
          <p>{kittyId}</p>
          <h5>Genes</h5>
          <p>{genes}</p>
          <h5>Generation</h5>
          <p>{generation}</p>
          <h5>Birth Time</h5>
          <p>{formattedBday}</p>
        </div>
        <div className="kitty-img">
          <img
            src={`https://img.cryptokitties.co/0x06012c8cf97bead5deae237070f9587f8e7a266d/${kittyId}.svg`}
            alt="kitty"
          />
        </div>
      </Card.Body>
    </Card>
  );
};
