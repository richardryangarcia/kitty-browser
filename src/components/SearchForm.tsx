import * as React from "react";
import { Form, Button, FormControlProps } from "react-bootstrap";

type SearchFormProps = {
  kittyId: string;
  handleSubmit: (event: React.FormEvent<FormControlProps>) => void;
  handleChange: (event: React.FormEvent<FormControlProps>) => void;
  handleRandomKitty: (event: React.FormEvent<FormControlProps>) => void;
};
export const SearchForm: React.FC<SearchFormProps> = ({
  handleChange,
  handleRandomKitty,
  handleSubmit,
  kittyId
}) => {
  return (
    <Form onSubmit={handleSubmit}>
      <Form.Control
        size="lg"
        type="text"
        placeholder="Kitty ID"
        value={kittyId}
        onChange={handleChange}
      />

      <div className="search-btns">
        <Button className="btn-search" variant="primary" type="submit">
          Search Kitty
        </Button>

        <Button
          className="btn-random"
          variant="success"
          onClick={(event: React.FormEvent<FormControlProps>) => {
            handleRandomKitty(event);
          }}
        >
          Random Kitty
        </Button>
      </div>
    </Form>
  );
};
