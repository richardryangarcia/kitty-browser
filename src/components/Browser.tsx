import React, { Component } from "react";
import { object } from "prop-types";
import Web3 from "web3";
import KittyCoreABI from "../contracts/KittyCoreABI.json";
import { AbiItem, AbiType } from "web3-utils";
import { CONTRACT_NAME, CONTRACT_ADDRESS } from "../config";
import { Container, Row, Col, FormControlProps } from "react-bootstrap";
import { KittyCard } from "./KittyCard";
import { SearchForm } from "../components/SearchForm";

type BrowserState = {
  kittyId: string;
  birthTime?: Date;
  generation: number;
  genes: number;
};

declare global {
  interface Window {
    web3: any;
  }
}

enum Mutability {
  View = "view",
  Nonpayable = "nonpayable",
  Payable = "payable",
  Pure = "pure"
}

class Browser extends Component<{}, BrowserState> {
  contracts: any;
  public static contextTypes = {
    drizzle: object
  };

  constructor(props: any, context: any) {
    super(props);
    this.state = {
      kittyId: "",
      birthTime: undefined,
      generation: 0,
      genes: 0
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleRandomKitty = this.handleRandomKitty.bind(this);
    this.mutability = this.mutability.bind(this);
    this.contracts = context.drizzle.contracts;
  }

  componentDidMount() {
    const web3 = new Web3(window.web3.currentProvider);
    // Initialize the contract instance
    const typedABI: AbiItem[] = KittyCoreABI.map(sig => {
      const type: AbiType = sig.type as AbiType;
      const item: AbiItem = {
        anonymous: undefined,
        constant: sig.constant,
        inputs: sig.inputs,
        name: sig.name,
        outputs: sig.outputs,
        payable: sig.payable,
        stateMutability: this.mutability(sig.stateMutability),
        type: type
      };
      return item;
    });

    const kittyContract = new web3.eth.Contract(
      typedABI, // import the contracts's ABI and use it here
      CONTRACT_ADDRESS
    );
    // Add the contract to the drizzle store
    this.context.drizzle.addContract({
      contractName: CONTRACT_NAME,
      web3Contract: kittyContract
    });
  }

  handleChange(event: React.FormEvent<FormControlProps>) {
    const value = event.currentTarget.value ? event.currentTarget.value : "";
    this.setState({
      kittyId: value,
      birthTime: undefined,
      generation: 0,
      genes: 0
    });
  }

  mutability(mutability: string | undefined) {
    switch (mutability) {
      case "view":
        return Mutability.View;
      case "nonpayable":
        return Mutability.Nonpayable;
      case "payable":
        return Mutability.Payable;
      case "pure":
        return Mutability.Pure;
      default:
        return undefined;
    }
  }

  async handleSubmit(event: React.FormEvent<FormControlProps>) {
    event && event.preventDefault();
    let response;
    try {
      response = await this.contracts.CryptoKitties.methods
        .getKitty(this.state.kittyId)
        .call();
      this.setState({
        birthTime: response.birthTime,
        generation: response.generation,
        genes: response.genes
      });
    } catch (e) {
      this.setState({
        birthTime: undefined,
        generation: 0,
        genes: 0
      });
    }
  }

  handleRandomKitty(event: React.FormEvent<FormControlProps>) {
    this.setState({
      kittyId: Math.round(Math.random() * 1713872).toString()
    });
    this.handleSubmit(event);
  }

  render() {
    const { birthTime, genes, generation, kittyId } = this.state;

    return (
      <div className="browser">
        <Container>
          <Row>
            <Col>
              <h1 className="header">Kitty Browser</h1>
              <SearchForm
                handleSubmit={this.handleSubmit}
                handleChange={this.handleChange}
                handleRandomKitty={this.handleRandomKitty}
                kittyId={kittyId}
              />
            </Col>
          </Row>
          <Row>
            <Col>
              {birthTime && (
                <KittyCard
                  birthTime={birthTime}
                  genes={genes}
                  generation={generation}
                  kittyId={kittyId}
                />
              )}
            </Col>
          </Row>
        </Container>
      </div>
    );
  }
}

export default Browser;
