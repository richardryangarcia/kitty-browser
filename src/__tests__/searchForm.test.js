import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { SearchForm } from "../components/SearchForm";
import { get } from "http";

describe("<SearchForm />", () => {
  let kittyId, handleSubmit, handleChange, handleRandomKitty;
  beforeEach(() => {
    kittyId = "470916";
    handleSubmit = () => {};
    handleChange = () => {};
    handleRandomKitty = () => {};
  });
  test("renders correctly", () => {
    const { asFragment } = render(
      <SearchForm
        kittyId={kittyId}
        handleSubmit={handleSubmit}
        handleChange={handleChange}
        handleRandomKitty={handleRandomKitty}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render a form with search and random button", () => {
    const { container, get } = render(
      <SearchForm
        kittyId={kittyId}
        handleSubmit={handleSubmit}
        handleChange={handleChange}
        handleRandomKitty={handleRandomKitty}
      />
    );
    const form = container.querySelector("form");
    const searchButton = container.querySelector(".btn-search");
    const randomButton = container.querySelector(".btn-random");
    expect(form).not.toBe(null);
    expect(searchButton).not.toBe(null);
    expect(randomButton).not.toBe(null);
  });
});
