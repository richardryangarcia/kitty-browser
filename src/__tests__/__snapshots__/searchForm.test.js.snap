// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm /> renders correctly 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <input
      class="form-control form-control-lg"
      placeholder="Kitty ID"
      type="text"
      value="470916"
    />
    <div
      class="search-btns"
    >
      <button
        class="btn-search btn btn-primary"
        type="submit"
      >
        Search Kitty
      </button>
      <button
        class="btn-random btn btn-success"
        type="button"
      >
        Random Kitty
      </button>
    </div>
  </form>
</DocumentFragment>
`;
